examples/cpp/src/dynamic_batching_example.cu

```diff
@@ -0,0 +1,279 @@
+
+  static void completion_callback(void *ptr) {
+    auto *promise = reinterpret_cast<std::promise<void> *>(ptr);
+    promise->set_value();
+    delete promise;
+  }
+};
+
+void dynamic_batching_example(
+    raft::resources const &res,
+    raft::device_matrix_view<const float, int64_t> dataset,
+    raft::device_matrix_view<const float, int64_t> queries) {
+  using namespace cuvs::neighbors;
+
+  // Number of neighbors to search
+  int64_t topk = 100;
+
+  // Streaming scenario: maximum number of requests in-flight
+  constexpr int64_t kMaxJobs = 1000;
+  // Streaming scenario: number of concurrent CUDA streams
+  constexpr int64_t kNumWorkerStreams = 5;
+
+  // Split the queries into two subsets to run every experiment twice and thus
+  // surface any initialization overheads.
+  int64_t n_queries_a = queries.extent(0) / 2;
+  int64_t n_queries_b = queries.extent(0) - n_queries_a;
+
+  auto queries_a = slice_matrix(queries, 0, n_queries_a);
+  auto queries_b = slice_matrix(queries, n_queries_a, n_queries_b);
+
+  // create output arrays
+  auto neighbors =
+      raft::make_device_matrix<uint32_t>(res, queries.extent(0), topk);
+  auto distances =
+      raft::make_device_matrix<float>(res, queries.extent(0), topk);
+  // slice them same as queries
+  auto neighbors_a = slice_matrix(neighbors, 0, n_queries_a);
+  auto distances_a = slice_matrix(distances, 0, n_queries_a);
+  auto neighbors_b = slice_matrix(neighbors, n_queries_a, n_queries_b);
+  auto distances_b = slice_matrix(distances, n_queries_a, n_queries_b);
+
+  // use default index parameters
+  cagra::index_params orig_index_params;
+
+  std::cout << "Building CAGRA index (search graph)" << std::endl;
+  auto orig_index = cagra::build(res, orig_index_params, dataset);
+
+  std::cout << "CAGRA index has " << orig_index.size() << " vectors"
+            << std::endl;
+  std::cout << "CAGRA graph has degree " << orig_index.graph_degree()
+            << ", graph size [" << orig_index.graph().extent(0) << ", "
+            << orig_index.graph().extent(1) << "]" << std::endl;
+
+  // use default search parameters
+  cagra::search_params orig_search_params;
+  // get a decent recall by increasing the internal topk list
+  orig_search_params.itopk_size = 512;
+  orig_search_params.algo = cagra::search_algo::SINGLE_CTA;
+
+  // Dynamic batching index takes the upstream index as a build parameter
+  // ("building" dynamic_batching::index means just wrapping the upstream and
+  // preparing IO buffers)
+  dynamic_batching::index_params<decltype(orig_index)> dynb_index_params{
+      cuvs::distance::DistanceType::L2Expanded, 2.0, orig_index,
+      /* We need to know the search parameters in advance for batching */
+      orig_search_params, nullptr, topk, queries.extent(1),
+      /* Configure the number and the size of IO buffers */
+      64, kNumWorkerStreams};
+
+  // "build" the index
+  dynamic_batching::index<float, uint32_t> dynb_index(res, dynb_index_params);
```

Comment:
Right now `dynb_index_params` holds a reference to the original index, and the new dynamic index is constructed using this params struct. Alternatively, one could pass the `orig_index` to the constructor of the dynamic index:

```c++
dynamic_batching::index_params<decltype(orig_index)> dynb_index_params{
      orig_search_params, nullptr, topk, queries.extent(1),
      /* Configure the number and the size of IO buffers */
      64, kNumWorkerStreams};

  // "build" the index
  dynamic_batching::index<float, uint32_t> dynb_index(res, dynb_index_params, orig_index);
```

I would slightly prefer this, because in my mind the `orig_index` is not a parameter, but the object around which we are adding a wrapper to do the dynamic batching (the same way as `dataset` is not a parameter of `index_params`). 

Let's wait for other opinions of this topic.

Reply:
I very much agree with this. I find it also particularly clumsy to use the aggregate constructor in the index params (in benchmarks and in the example file) and a little unsafe to keep the reference there (a user must keep the original index alive).

In here, I attempted to keep the same build API as the other indexes, but failed (`::build(...)` function arguments don't fit here at all). I'd be happy to hear further suggestions and adapt the code once we have a consensus.

Reply:
I updated the API to pass more parameters via the index constructor and made the index parameters non-template.